### PR TITLE
refactor: deploy the training-operator with kubernetes resources

### PR DESCRIPTION
**Pull request by DnPlas**
This commit refactors the way the training-operator is deployed, as instead of using a sidecar container that runs the workload, we are now applying the Deployment and all the Kubernetes resources required by the training-operator controller to be able to mange training resources.
We are introducing this change in preparation for the upcoming 1.8 version, as it introduces the hard dependency on a Kubernetes Secret mounted in a volume for the training-operator workload to start. For more details please refer to canonical/training-operator#159.

### CODE REVIEW

**Review by orfeas-k — COMMENTED** [review]
left some minor comments and I 'm testing now

File: src/charm.py [hunk 1]
```diff
@@ -38,6 +38,28 @@ def __init__(self, *args):
...
+        self._lightkube_field_manager = "lightkube"
+        self._context = {
+            "namespace": self._namespace,
+            "app_name": self._name,
+            "training_operator_image": self._image,
+            "metrics_port": METRICS_PORT,
+        }
+
+        self._k8s_resource_handler = None
+        self._crd_resource_handler = None
```
[the remark this answers is not in the document]
Comment (DnPlas) [reply src/charm.py:53]: I guess you mean the 

```python
self._k8s_resource_handler = None
self._crd_resource_handler = None
```
This is just left as None to allow us to create a handler with [this](https://github.com/canonical/training-operator/blob/main/src/charm.py#L86) property.

btw, they are in [main](https://github.com/canonical/training-operator/blob/main/src/charm.py#L66C1-L67C42) as well, this PR just re-order them.

File: src/charm.py [hunk 2]
```diff
@@ -114,36 +114,9 @@ def crd_resource_handler(self):
...
-        return Layer(layer_config)
-
-    def _check_container_connection(self):
-        """Check if connection can be made with container."""
-        if not self.container.can_connect():
-            raise ErrorWithStatus("Pod startup is not complete", MaintenanceStatus)
-
     def _check_leader(self):
         """Check if this unit is a leader."""
+        self.framework.observe(self.on.update_status, self._on_event)
```
[the remark this answers is not in the document]
Comment (DnPlas) [reply src/charm.py:119]: Mmmh, I think this is a typo, not sure how this end up here.

[the remark this answers is not in the document]
Comment (DnPlas) [reply src/charm.py:119]: I have removed it.

File: src/templates/rbac_manifests.yaml.j2 [hunk 3]
```diff
@@ -1,9 +1,23 @@
...
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  labels:
+    app: {{ app_name }}
+  name: {{ app_name }}
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: training-operator
```
[the remark this answers is not in the document]
Comment (DnPlas) [reply src/templates/rbac_manifests.yaml.j2:11]: Updated.

**Review by orfeas-k — APPROVED** [review]

**Comment by DnPlas** [comment]
@orfeas-k COS integration tests were failing, this PR may need a second review.

**Review by orfeas-k — APPROVED** [review]

**Pull request merged**